addons/stock_account/models/stock_valuation_layer.py

```diff
@@ -33,13 +33,24 @@ class StockValuationLayer(models.Model):
     account_move_line_id = fields.Many2one('account.move.line', 'Invoice Line', readonly=True, check_company=True, index="btree_not_null")
     reference = fields.Char(related='stock_move_id.reference')
     price_diff_value = fields.Float('Invoice value correction with invoice currency')
+    warehouse_id = fields.Many2one('stock.warehouse', string="Warehouse IN", compute='_compute_warehouse_id', search='_search_warehouse_id')
 
     def init(self):
         tools.create_index(
             self._cr, 'stock_valuation_layer_index',
             self._table, ['product_id', 'remaining_qty', 'stock_move_id', 'company_id', 'create_date']
         )
 
+    def _compute_warehouse_id(self):
+        for record in self:
+            if record.stock_move_id.location_id.usage == "internal":
+                record.warehouse_id = record.stock_move_id.location_id.warehouse_id.id
+            else:
+                record.warehouse_id = record.stock_move_id.location_dest_id.warehouse_id.id
+
+    def _search_warehouse_id(self, operator, value):
+        return [('id', operator, value)]
```

Comment:
This doesn't work. Just use related field as the spec says.

Reply:
Hello @yhu-odoo we can't make this field related, cause its value comes based on condition. that's why I make it compute and search.